### PR TITLE
Refactor/deployerlib

### PR DESCRIPTION
**Pull request by schrodit**
**How to categorize this PR?**

/area deployers
/priority 3

**What this PR does / why we need it**:
- added documentation for the deployer configuration and deployer helm charts
- added labels as additional selection method
- added integration tests for target selection

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```doc operator
Documentation for the deployer controller configuration has been added.
In addition the basic target selector is explained in the common deployer docs.
```
```feature operator
It is now possible to also use labels to select targets for deployers.
```

### CODE REVIEW

File: charts/container-deployer/values.yaml [hunk 1]
```diff
@@ -10,12 +10,15 @@ deployer:
 #  namespace: ""
   initContainer:
     repository: eu.gcr.io/gardener-project/landscaper/container-deployer-init
+#    tag: "latest"
   waitContainer:
     repository: eu.gcr.io/gardener-project/landscaper/container-deployer-wait
-  defaultImage: {}
+#    tag: "latest"
+#  defaultImage:
```
Comment (MrBatschner) [inline charts/container-deployer/values.yaml:17]: `initContainer` and `waitContainer` have dedicated `respository` and `tag` fields. Shouldn't we have the same for `defaultImage`? Just for consistency? Of course that would require changing the `_helpers.tpl` as well...

File: apis/deployer/mock/v1alpha1/types_config.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	lsv1alpha1 "github.com/gardener/landscaper/apis/core/v1alpha1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Configuration is the helm deployer configuration that configures the controller
+type Configuration struct {
+	metav1.TypeMeta `json:",inline"`
+	// OCI configures the oci client of the controller
+	OCI *config.OCIConfiguration `json:"oci,omitempty"`
```
Comment (MrBatschner) [inline apis/deployer/mock/v1alpha1/types_config.go:20]: Does the mock-deployer even need an OCI configuration?

**Review by MrBatschner — COMMENTED** [review]
Looks good so far. I found some minor things which I commented inline.

File: docs/deployer/container.md [hunk 3]
```diff
@@ -123,3 +129,132 @@ perform specific actions.
...
+  
+## Deployer Configuration
+
+When deploying the container deployer controller it can be configured using the `--config` flag and providing a configuration file.
+
+The structure of the provided configuration file is defined as follows.
+
+:warning: Keep in mind that when deploying with the helm chart the configuration is abstracted using the helm values. See the [helm values docs](#helm-chart-configuration) for details when deploying with the helm chart.
+```yaml
+apiVersion: container.deployer.landscaper.gardener.cloud/v1alpha1
```
[the remark this answers is not in the document]
Comment (schrodit) [reply docs/deployer/container.md:141]: yes make sense to just link them and keep the comments in the values.yaml only 👍🏻

File: charts/container-deployer/values.yaml (hunk 1 above)
Comment (MrBatschner) [reply charts/container-deployer/values.yaml:17]: **Definitely needs a change in `_helpers.tpl`** which made the integration test fail. If `defaultImage` is commented out like in here,
```
{{- if .Values.deployer.defaultImage.image }}
defaultImage:
  image: "{{ .Values.deployer.defaultImage.image }}"
{{- end }}
```
in https://github.com/gardener/landscaper/blob/a1e3520394e8e244816e0c7d5c3c581d2c6c3a20/charts/container-deployer/templates/_helpers.tpl#L76 will `nil`.

**Comment by schrodit** [comment]
@MrBatschner PTAL

I will also check why the integration tests are failing. look slike an issue in the testmachinery

**Review by MrBatschner — APPROVED** [review]
/lgtm

**Pull request merged**